### ISSUE TITLE
Deny HTTP endpoints for backup (Minio)

### ISSUE THREAD

**Issue by oskapt** [longhorn/longhorn#705]
On the back of #704 we also agreed that if we can specify a trusted CA for the backup endpoint, there is no reason not to enable TLS for Minio. That will enable Longhorn to safely deny any HTTP endpoint for backups, which would be a Good Thing:tm:.

**Comment by yasker** [comment]
I am not sure it's necessary since I don't think you need to require HTTPS for internal networks or the testing setup. Sounds like the choice is up to the user.

**Comment by oskapt** [comment]
bump. 

if you don't want to do it, please close the issue.

**Comment by yasker** [comment]
@oskapt Just to be clear, it's not about I want to do it or not. As stated above, my argument is it should be up to the user. I don't think it's needed for internal networks or the testing setup. S3 endpoint is not necessarily always on the internet. And if we require SSL for S3, how about NFS? I wasn't involved in the original discussion so I am not sure about the context. I am open to discuss more here.

**Comment by oskapt** [comment]
Even on internal networks, encryption is good. If we're streaming backups of storage data across a network, someone could still intercept and read that data. I think the argument is that if given the choice to encrypt or not encrypt, it's always better to encrypt. If we allow people to stream data to an HTTP endpoint, then at some point someone will stream important data to that endpoint, and if that results in any sort of compromise, we could have avoided it. 

Wrapping things in TLS is trivial and costs nothing. Even with Minio it's easy to do it. We don't allow Rancher to run backups to an HTTP endpoint, so Longhorn would be consistent in enforcing the same.

**Comment by yasker** [comment]
We will also add end-to-end encryption in #902.

If we're going to require SSL, we basically have to support a self-signed certificate (#704) and change our Minio demo pod. But I am not sure if #704 is needed. You can easily get a CA-signed certificate from LetsEncrypt, and Longhorn don't need to do anything else to support it. But if you want to use one specific self-signed certificate, we basically have to find a way to distribute that certificate to every node, ready for the S3 usage. And when the user changes the certificate (e.g. certification expired), we need to update everything as well. All of those can be avoided by just using a free CA-signed certificate.

But of course, we can require SSL and package a LetsEncrypt certificate in our Minio demo pod. Though how about NFS? We don't have encryption for NFS in this case. I guess we can say it's case by case, but it's less ideal than implementing #902 . I am on the fence right now. We can keep this issue open and see what others think.

**Comment by oskapt** [comment]
There's an inconsistency w/ using the same Minio endpoint for Rancher and Longhorn. Rancher will accept a cert from an unknown authority if you add the CA. Longhorn won't. That means you can't use the same Minio environment for backups for both. Saying, "oh you can just use LetsEncrypt" forces the user to use LetsEncrypt. What if they don't want to or can't? What about organizations that run their own internal CA?

My network sits behind an edge NAT and then my ISP backhauls all the traffic to an egress point in Santiago, where it goes through another NAT. It's not possible for LE to use HTTP to validate my requests, and setting up wildcard DNS is more headache than it's worth for Minio and Longhorn. I run an internal CA, and while that works for Rancher, I'm unable to make backups with Longhorn.

#704 would fix this.

This issue is about removing HTTP as an option, and I think it comes down to the message that we want to send to our users. Is security important? Is encryption important? Do we default to being secure and encrypted, or do we default to being insecure and unencrypted?

**Comment by azbpa** [comment]
> There's an inconsistency w/ using the same Minio endpoint for Rancher and Longhorn. Rancher will accept a cert from an unknown authority if you add the CA. Longhorn won't. That means you can't use the same Minio environment for backups for both. Saying, "oh you can just use LetsEncrypt" forces the user to use LetsEncrypt. What if they don't want to or can't? What about organizations that run their own internal CA?

Facing the same issue while playing around with Longhorn. Think security is a key feature, even in internal networks. In our case Let's Encrypt isn't an option as well and we are using internal PKI for signing every certificate. So also for an internal Minio installation.

So please add the option to add custom certificate authorities for the backup endpoints like it is possible from Rancher server.

**Comment by yasker** [comment]
@azbpa Custom CA will be in the GA release, and it's tracked at #704

**Comment by oskapt** [comment]
Just for the record, bots that close valid issues because the maintainer/org doesn't care about the issue are terrible. Imagine if I worked for a company that assigned me tasks that I ignored, so after a while a bot came around and removed those tasks from my inbox and gave me a paycheck instead. Does such a company exist or stay in business? Granted, this issue is almost five years old and might already be resolved...but who knows? Let's just close it and pretend in never happened, mmmkay???
